### PR TITLE
add option to encapsulate

### LINKED ISSUE #197: Introduce an option to encapsulate the plugin
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the feature has not already been requested

### 🚀 Feature Proposal

Introduce an option `encapsulate?: boolean`. Defaults to `false`. When set to `true`, keep the plugin encapsulated.

### Motivation

`fastify-plugin` currently makes all plugins accessible by setting `skip-override`. It would be nice to have an option to keep it encapsulated, so you can set a plugin's name, metadata, etc without affecting the plugin's scope. 

Would you consider a PR for this?

### Example

```js
module.exports = fp(plugin, {
  fastify: '4.x',
  decorators: {
    fastify: ['plugin1', 'plugin2'],
    reply: ['compress']
  },
  dependencies: ['plugin1-name', 'plugin2-name'],
  encapsulate: true,
})
```

### PR DESCRIPTION
**Pull request by dwickern**
#### Checklist

- [ ] run `npm run test` and `npm run benchmark`
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

Resolves #197

### CODE REVIEW

File: README.md
Comment (mcollina) [inline README.md:100]: Maybe link to the fastify docs about encapsulation?

File: plugin.js
Comment (jsumners) [inline plugin.js:44]:
```suggestion
  fn[Symbol.for('skip-override')] = options?.encapsulate ?? true
```

**Comment by dwickern** [comment]
Will this change interact safely with [fastify-autoload's default with the same name](https://github.com/fastify/fastify-autoload/blob/697c9d9840bb224c3498a30390c6c7937796386e/index.js#L19)?

File: plugin.js
Comment (dwickern) [reply plugin.js:44]: A few lines above we make sure `options` is a non-null object

Comment (jsumners) [reply plugin.js:44]: I would still rather not rely on a difficult to interpret coercive `!`.

Comment (Uzlopak) [reply plugin.js:44]: If we check before that options is a POJO then we can avoid the question mark. But regarding the coercion with an exclamation mark i agree with jsumners

```suggestion
  fn[Symbol.for('skip-override')] = options.encapsulate ?? true 
```

File: test/test.js
Comment (Eomm) [inline test/test.js:269]:
```suggestion
    t.error(err)
```

File: README.md
Comment (Eomm) [inline README.md:100]:
```suggestion

By default, `fastify-plugin` breaks the [encapsulation](https://github.com/fastify/fastify/blob/HEAD/docs/Reference/Encapsulation.md) but you can optionally keep the plugin encapsulated.
```

File: README.md
Comment (Eomm) [inline README.md:102]:
```suggestion

```

File: test/test.js
Comment (Eomm) [inline test/test.js:265]: We need a test that verifies what the docs say: does it run all the validations?

Comment (dwickern) [reply test/test.js:265]: Would you say the 1 test is enough? I'd rather avoid repeating the whole test suite

Comment (Eomm) [reply test/test.js:265]: All the cases that have not been tested have an undetermined state, so the maintainers can't help whoever opens an issue.

In this case, the maintainer should add a test case for it and add it to the repository etc.. in an endless pain loop.

So, we can discuss about the minimal test suite we need, but yes: 1 test is not enough of a HUGE feature like this.

So I think the minimal test suite for this test suite is:

- check that the feature works as expected:
  - [x] are decorators encapsulated?
  - [ ] is the fastify version check working?
  - [ ] are mandatory decorators checks applied?
  - [ ] is the required plugins option working?
- verify how fastify reacts to this configuration (plugin name chaining)

**Review by mcollina — APPROVED** [review]
lgtm

**Review by Eomm — CHANGES_REQUESTED** [review]
Could you integrate this https://github.com/fastify/fastify-plugin/commit/00d2bdd015a97c8b7f81b5598e1bfc8d46365cc3
commit into this PR before landing?

**Comment by dwickern** [comment]
Thanks @Eomm; I pushed your changes to this branch

**Review by Eomm — APPROVED** [review]

**Pull request merged**
